### PR TITLE
feat: Highlight exit and yield points

### LINKED ISSUE #4691: Semantic highlighting for function exit points
[As seen in IntelliJ](https://intellij-rust.github.io/2020/06/01/changelog-123.html)

### PR DESCRIPTION
**Pull request by Veykril**
hvx3IWepc5
Code_YyMhqES0LX

Fixes #4691

### CODE REVIEW

**Comment by matklad** [comment]
Moves in the right direction! Couple of relevant thoughts:

* a related functionality is highlighting all exit points of the function, but we probably want a separate request for that (and support in client)
* document highlights is less than ideal name here. Perhaps just `highlight_references`?

**Comment by kjeremy** [comment]
>     * a related functionality is highlighting all exit points of the function, but we probably want a separate request for that (and support in client)

I don't think it needs to be a separate request See https://github.com/microsoft/language-server-protocol/issues/1047

**Comment by Veykril** [comment]
Ye I think we don't need a separate request for that, but then the question is how we highlight exit points in regards to the trailing expression as that can overlap with the general ident highlight.

**Comment by matklad** [comment]
If the cursor is on ? / return or return type (`-> Ty`) we should highlight exit points, including expressions in the tail position.

if the cursor is on async / await, we should highlight all yield points.

if the cursor is within trailing expression, we should highlight references or nothing.


The name for all three I think is “highlight_related”?

**Comment by Veykril** [comment]
Yield point and exit point highlighting is implemented as well now. Gotta fix(and probably find) a few more special cases as well as do some minor cleanup still

**Comment by Veykril** [comment]
Really odd, so VSCode refuses to show highlighting when hovering over some exit points but not on others even though it works fine for them in tests...

**Comment by Veykril** [comment]
Okay things are working ***but*** VSCode doesn't send a highlight request when the cursor goes onto non identifier symbols from what I can tell, that means clicking on `?` and `->` doesn't send a request currently. I suppose this is a bug on VSCode's side as I don't see anything in this regard in the lsp docs.

**Comment by Veykril** [comment]
Currently when the trailing expression is an `if` or `match` expression the entire expression gets highlighted as an exit point. Now I wonder if it would make more sense to only highlight the arm/branch expressions in this case? Seems more elegant to me.

**Comment by matklad** [comment]
Yeah, it’s better to highlight expressions in the tail position. I think we should have functionality for that implemented somewhere already, but I don’t remember where. Ok-wrapping maybe?

File: crates/ide/src/highlight_related.rs [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+//
+// Highlights constructs related to the thing under the cursor:
+// - if on an identifier, highlights all references to that identifier in the current file
+// - if on an `async` or `await token, highlights all yield points for that async context
+// - if on a `return` token, `?` character or `->` return type arrow, highlights all exit points for that context
+pub(crate) fn highlight_related(
+    sema: &Semantics<RootDatabase>,
+    position: FilePosition,
+) -> Option<Vec<DocumentHighlight>> {
+    let _p = profile::span("document_highlight");
```
Comment (matklad) [inline crates/ide/src/highlight_related.rs:32]:
```suggestion
    let _p = profile::span("highlight_related");
```

File: crates/ide/src/highlight_related.rs [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+    let syntax = sema.parse(position.file_id).syntax().clone();
+
+    let token = pick_best_token(syntax.token_at_offset(position.offset), |kind| match kind {
+        QUESTION => 2, // prefer `?` when the cursor is sandwiched like `await$0?`
```
Comment (matklad) [inline crates/ide/src/highlight_related.rs:36]:
```suggestion
        T![?] => 2, // prefer `?` when the cursor is sandwiched like `await$0?`
```

ditto for other tokens

**Comment by matklad** [comment]
Got it! We have “change return type to Result” assist!

File: crates/ide/src/highlight_related.rs [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+};
+use syntax::{
+    ast, match_ast, AstNode,
+    SyntaxKind::{ASYNC_KW, AWAIT_KW, QUESTION, RETURN_KW, THIN_ARROW},
+    SyntaxNode, SyntaxToken, TextRange, WalkEvent,
+};
+
+use crate::{display::TryToNav, references, NavigationTarget};
+
+pub struct DocumentHighlight {
```
Comment (matklad) [inline crates/ide/src/highlight_related.rs:17]: Less than ideal name, as we don’t use document terminology in ide. I guess just RelatedRange or HighlightedRange would work?

**Comment by Veykril** [comment]
Oh ye looks like that assist does similar things to what is wanted here, I'll take a look at that tomorrow

File: crates/ide/src/highlight_related.rs [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+                break 5;
+             // ^^^^^^^
+            }
+            8 => 'a: loop {
+                'b: loop {
+                    break 'a 5;
+                 // ^^^^^^^^^^
+                    break 'b 5;
+                    break 5;
+                };
```
Comment (Veykril) [inline crates/ide/src/highlight_related.rs:575]: I wonder if I went a bit over the top with this to also properly consider (labeled) breaks in loops 🤔

**Comment by Veykril** [comment]
Managed to simplify things a bit more, but I couldn't really reuse the `wrap_return_type_in_result` logic for tails fully as its a bit different than whats needed here. Might be able to merge the logic somehow still but I'll take a look at that another time.
bors r+

**Pull request merged**